google/cloud/pubsublite/publisher_connection.cc

```diff
@@ -0,0 +1,215 @@
+  Value lang;
+  *lang.mutable_string_value() = "CPP";
+  metadata_map["language"] = std::move(lang);
+  Value framework_val;
+  *framework_val.mutable_string_value() = framework;
+  metadata_map["framework"] = framework_val;
+  Value minor_version;
+  minor_version.set_number_value(version_minor());
+  Value major_version;
+  major_version.set_number_value(version_major());
+  std::string serialized_context;
+  context.SerializePartialToString(&serialized_context);
+  Base64Encoder encoder;
+  for (char const c : serialized_context) encoder.PushBack(c);
+  return std::move(encoder).FlushAndPad();
+}
+
+ClientMetadata MakeClientMetadata(Topic const& topic, std::uint32_t partition) {
+  ClientMetadata metadata;
+  metadata["x-goog-request-params"] =
+      absl::StrCat("partition=", partition, "&", "topic=", topic.FullName());
+  metadata["x-goog-pubsub-context"] = GetSerializedContext("CLOUD_PUBSUB_SHIM");
+  return metadata;
+}
+
+StatusOr<std::unique_ptr<PublisherConnection>> MakePublisherConnection(
+    Topic topic, Options opts) {
+  if (!opts.has<GrpcNumChannelsOption>()) {
+    // Default to 20 channels per client to avoid limitations on streams and
+    // requests per-channel.
+    opts.set<GrpcNumChannelsOption>(20);
+  }
+  if (!opts.has<EndpointOption>()) {
+    auto endpoint = GetEndpoint(topic.location());
+    if (!endpoint) {
+      return Status{StatusCode::kInvalidArgument, "`topic` not valid"};
+    }
+    opts.set<EndpointOption>(*std::move(endpoint));
+  }
+  std::unique_ptr<BackgroundThreads> background_threads =
+      MakeBackgroundThreadsFactory(opts)();
+  CompletionQueue cq = background_threads->cq();
+  std::shared_ptr<BackoffPolicy const> backoff_policy =
+      std::make_shared<ExponentialBackoffPolicy>(std::chrono::milliseconds{10},
+                                                 std::chrono::seconds{10}, 2.0);
+  AsyncSleeper sleeper = [cq](std::chrono::milliseconds dur) mutable {
+    return cq.MakeRelativeTimer(dur).then(
+        [](future<StatusOr<std::chrono::system_clock::time_point>> f) {
+          auto status = f.get();
+          if (!status.ok()) {
+            GCP_LOG(INFO) << "`MakeRelativeTimer` returned a non-ok status: "
+                          << status.status();
+          };
+        });
+  };
+
+  auto publisher_service_stub = CreateDefaultPublisherServiceStub(cq, opts);
+  auto batching_options = CreateBatchingOptions(opts);
+
+  std::function<std::shared_ptr<
+      Publisher<google::cloud::pubsublite::v1::Cursor>>(std::uint32_t)>
+      partition_publisher_factory = [topic, backoff_policy, batching_options,
+                                     cq, publisher_service_stub,
+                                     sleeper](std::uint32_t partition) {
+        InitialPublishRequest ipr;
+        *ipr.mutable_topic() = topic.FullName();
+        ipr.set_partition(partition);
+        AlarmRegistryImpl alarm_registry{cq};
+        auto stream_factory = MakeStreamFactory(
```

Comment:
put this on line 162, outside the lambda

Reply:
`auto stream_factory = MakeStreamFactory(publisher_service_stub, cq, MakeClientMetadata(topic, partition));` requires `partition` which is a parameter of the lambda

Reply:
you're right, sorry